pypesto/sampling/geweke_test.py

```diff
@@ -0,0 +1,204 @@
+import numpy as np
+from scipy.stats import norm
+
+
+def spectrum(x: np.array,
+             nfft: int = None,
+             nw: int = None):
+    """
+    Power spectral density using Hanning window.
+
+    Parameters
+    ----------
+    x:
+        Fraction/fragment of one single parameter
+        of the MCMC chain.
+    nfft:
+        The n-point discrete Fourier Transform.
+    nw:
+        Number of windows.
+
+    Returns
+    -------
+    spectral_density:
+        The spectral density.
+
+    """
+
+    if nfft is None:
+        nfft = np.min(len(x), 256)
+
+    if nw is None:
+        nw = int(nfft / 4)
+
+    n_overlap = int(nw / 2)
+
+    # Hanning window
+    w = .5 * (1 - np.cos(
+        2 * np.pi * np.transpose(np.arange(1, nw+1)) /
+        (nw + 1)))
+    n = len(x)
+    if n < nw:
+        x[nw] = 0
+        n = nw
+
+    # Number of windows
+    k = int((n - n_overlap) / (nw - n_overlap))
+    index = np.arange(nw)
+    # Normalizing scale factor
+    kmu = k * np.linalg.norm(w) ** 2
+    spectral_density = np.zeros((nfft))
+
+    for i in range(k):
+        xw = np.multiply(w, x[index])
+        index += (nw - n_overlap)
+        Xx = np.absolute(np.fft.fft(xw, n=nfft, axis=0)) ** 2
+        spectral_density += Xx
+
+    # Normalize
+    spectral_density = spectral_density * (1 / kmu)
+
+    n2 = int(nfft / 2)
+    spectral_density = spectral_density[0:n2]
+
+    return spectral_density
+
+
+def spectrum0(x: np.array):
+    """
+    Calculates the spectral density at frequency zero.
+
+    Parameters
+    ----------
+    x:
+        Fraction/fragment of the MCMC chain.
+
+    Returns
+    -------
+    spectral_density_zero:
+        Spectral density at zero.
+
+    """
+
+    n_samples, n_par = x.shape
+    spectral_density_zero = np.zeros((1, n_par))
+
+    for i in range(n_par):
+        spectral_density_zero[:, i] = \
+            spectrum(x[:, i], n_samples)[0]
+    return spectral_density_zero
+
+
+def geweke_test(chain: np.array,
+                a: float = 0.1,
+                b: float = 0.5):
+    """
+    Performs a Geweke test on a chain using the first
+    "a" fraction and the last "b" fraction of it for
+    comparison. Test for equality of the means of the
+    first a% and last b% of a Markov chain.
+
+    See:
+    Stephen P. Brooks and Gareth O. Roberts.
+    Assessing convergence of Markov chain Monte Carlo
+    algorithms. Statistics and Computing, 8:319--335, 1998.
+
+    Parameters
+    ----------
+    chain
+    a:
+        First fraction of the MCMC chain. Default 10%.
+    b:
+        Second fraction of the MCMC chain. Default 50%.
+
+    Returns
+    -------
+    z_score:
+        Z-score of the Geweke test.
+    p:
+        Significance level of the Geweke test.
+
+    """
+
+    nsimu, _ = chain.shape
+
+    # Define First fraction
+    index_a = int(a * nsimu)
```

Comment:
had to keep the `int` statement since all these values are used for array indexing

Reply:
`math.floor` should work. reason: it is just conceptually not clear that `int()` floors values when converting.

Reply:
or `int(floot(.))`

Reply:
what about `np.floor(X).astype(int)`

Reply:
Seems overly complicated. :) `int()` is a built-in and very well defined... (and i am not aware of a language where this would do anything else than truncating...)

Reply:
ok fine I guess :laughing: one would need to look up in IEEE standards, but should be fine.